### DISCUSSION TITLE
Document exceeds maximum size allowed by Algolia

### DISCUSSION THREAD

**Discussion by wlfcss** [requarks/wiki#3675] [Error / Bug Report]
**Describe the bug**
an error toast：Document exceeds maximum size allowed by Algolia.

**To Reproduce**
1. chose the search Engine：Algolia
2. Fill in the APPID & Admin API Key  correctly
3. **Rebuild Index** then Save these settings
4. See error：Document exceeds maximum size allowed by Algolia

**Expected behavior**
I hope Algolia works，but is not, I read the docs about the Algolia,But this error seems to have nothing to do with my Alogia subscription plan，Maybe I got it wrong？

**Host Info (please complete the following information):**
 - local：OSX 10.15.3 - Chrome 81.0.4044.122
 - server-OS: [ubuntu18.04]
 - Wiki.js version: [latest 2.3.77]
 - Database engine: [postgres]
 - language: ZH-cn

**Additional context**
 - none

**Comment by fferraro87** [comment]
same problem here

**Comment by bmann** [comment]
Just set this up, getting it as well

**Comment by wlfcss** [comment]
> Just set this up, getting it as well

![image](https://user-images.githubusercontent.com/12138285/83115793-7cd93080-a0fd-11ea-9b79-bcf843886f73.png)

- doesn’t seem to be working

**Comment by JorgenSolli** [comment]
Having the same issue.

**Comment by jayzalowitz** [comment]
Same issue here.

**Comment by jayzalowitz** [comment]
https://github.com/Requarks/wiki/issues/2089

**Comment by jayzalowitz** [comment]
Describe the bug
When you have a large document this gets thrown and no documents get indexed:
https://github.com/Requarks/wiki/blob/master/server/modules/search/algolia/engine.js#L142

To Reproduce
Make a very large document

Expected behavior
At least some documents would get through

Additional context
https://www.algolia.com/doc/guides/sending-and-managing-data/prepare-your-data/how-to/indexing-long-documents/ splitting by line would work just as well for these documents. At the very least telling what document was causing the problem would be great while trying to rebuild index.

**Comment by jayzalowitz** [comment]
Skipping indexing of those docs instead of failing for all indexes would be a great temp fix.

**Comment by MisterRee** [comment]
Same issue. Would be great to get a fix for this, a wiki without the ability to efficiently search the data is useless.

**Comment by tobias-graf-p** [comment]
Same issue here. Would be great if this could be fixed soon.

**Comment by joppehoekstra** [comment]
Same issue here. A solution would be very much appreciated!

**Comment by lcalem** [comment]
I'm having the same issue and would probably agree with temporarily removing or splitting the large document while an "ignore larger files" option is implemented but I don't even know which document is too big. Maybe as a quick improvement including the name of the problematic document in the message would be helpful?

**Comment by winkelement** [comment]
Now i run into the same issue. There is no way for me to find a workaround as there is no easy way to tell which document is too large. Are there any plans to tackle this bug? Right now this renders the Algolia implementation non-functional for existing content if you happen to have a large document.

**Reply by NGPixel** [reply]
Algolia is the only search engine that impose such arbitrary low document size limits.

If someone wants to submit a PR that automatically splits documents over multiple records and handle the query part to avoid duplicates, I'll review it, but otherwise I suggest using an alternative search engine that doesn't have such absurd limitations.

**Reply by winkelement** [reply]
I understand. I switched to PostgreSQL search which seems sufficient for now. Thanks for the clarification.

**Comment by dukei** [comment]
@NGPixel Algolia has raised limitation for a record size up to 100kb: https://support.algolia.com/hc/en-us/articles/4406981897617-Is-there-a-size-limit-for-my-index-records-/

Can you please update the limit in source code. You just need to replace 10 to 100 here: https://github.com/Requarks/wiki/blob/f55caab359ca7715012e3f9a3b9e31f16cccf399/server/modules/search/algolia/engine.js#L128

I tried and it works.

**Comment by Aemony** [comment]
I am baffled this has been implemented this way, and that it has seemingly remained unfixed for at least two years. You're not even told which page is the cause of it in neither the UI nor the server logs...

"Oh, okay, but at least it creates new records when existing pages are saved?" you might ask, and you'd be *wrong.*

For existing pages, Wiki.js sends a request body that includes `partialUpdateObjectNoCreate` which basically tells Algolia to **not** create the record if it does not exist. This means that full rebuild in the settings that fails because a single page _dares_ to be larger than 10 KB ? Well, any existing content is _dependent_ on that to actually work as it is the only way to create records for all existing pages on Algolia... That is purely a limitation on Wiki.js and has nothing to do with any limitations (absurd or otherwise) on Algolia's end.

If you cannot fix the "max document size" when rebuilding the whole index somehow, the only alternative to populate Algolia with all existing wiki pages is to actually *delete* all existing wiki pages one by one and *recreate* them, as this will result in Wiki.js *finally* telling Algolia to create a new record for the page...

/rant

**Reply by NGPixel** [reply]
Algolia is literally the only service that imposes a ridiculously low limit on document size and suggest to create multiple entries for the same document to handle the full content (both of which are incredibly stupid). The current implementation isn't optimized for how Algolia works. It was quickly adapted from other search modules. Feel free to submit a PR with the "correct" implementation. Fixing/optimizing a search module for a service with such stupid restrictions is simply not a priority at the moment.
